IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,683 @@
+
+Error handling:
+* Generally this class has three error cases: success, recoverable
+  error, and unrecoverable error. Whenever possible, recoverable
+  errors should be handled here, locally, without surfacing them
+  to the caller. (For example, when writing an account to the
+  keychain, the SecItemUpdate() may fail since the item isn't in
+  the keychian yet. This is normal, and the code continues using
+  SecItemAdd() without returning an error. If this add failed, a
+  non-recoverable error would be returned to the caller.) Where
+  applicable, macOS keychain OSStatus results are surfaced to the
+  caller as MSID-standard NSError objects.
+
+Additional Notes:
+* For a given <access_group>, multiple credentials are stored in
+  a single Type 1 keychain item.  This is a work-around for a macOS
+  keychain limitation related to ACLs (Access Control Lists) and
+  is intended to minimize macOS keychain access prompts.  Once
+  an application has access to the keychain item it can generally
+  access and update credentials without further keychain prompts.
+
+Reference(s):
+* Apple Keychain Services: https://developer.apple.com/documentation/security/keychain_services?language=objc
+* Schema:
+https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=%2FUnifiedSchema%2FSchema.md&version=GBdev
+
+*/
+
+static NSString *s_defaultKeychainGroup = @"com.microsoft.identity.universalstorage";
+static NSString *s_defaultKeychainLabel = @"Microsoft Credentials";
+static MSIDMacKeychainTokenCache *s_defaultCache = nil;
+
+@interface MSIDMacKeychainTokenCache ()
+
+@property (readwrite, nonnull) NSString *keychainGroup;
+@property (readwrite, nonnull) NSDictionary *defaultAccountQuery;
+
+@end
+
+@implementation MSIDMacKeychainTokenCache
+
+#pragma mark - Public
+
++ (NSString *)defaultKeychainGroup
+{
+    return s_defaultKeychainGroup;
+}
+
+// Set the default keychain group
+//
+// Errors:
+// * @throw - attempt to change the default keychain group after being initialized
+//
++ (void)setDefaultKeychainGroup:(NSString *)defaultKeychainGroup
+{
+    MSID_TRACE;
+
+    if (s_defaultCache)
+    {
+        MSID_LOG_ERROR(nil, @"Failed to set default keychain group, default keychain cache has already been instantiated.");
+
+        @throw @"Attempting to change the keychain group once AuthenticationContexts have been created or the default keychain cache has been retrieved is invalid. The default keychain group should only be set once for the lifetime of an application.";
+    }
+
+    MSID_LOG_INFO(nil, @"Setting default keychain group.");
+    MSID_LOG_INFO_PII(nil, @"Setting default keychain group to %@", defaultKeychainGroup);
+
+    if ([defaultKeychainGroup isEqualToString:s_defaultKeychainGroup])
+    {
+        return;
+    }
+
+    if (!defaultKeychainGroup)
+    {
+        defaultKeychainGroup = [[NSBundle mainBundle] bundleIdentifier];
+    }
+
+    s_defaultKeychainGroup = [defaultKeychainGroup copy];
```

Comment:
you're changing a static variable here. Should you make it thread safe?

Reply:
I'll double-check this. The code is the same as on iOS.